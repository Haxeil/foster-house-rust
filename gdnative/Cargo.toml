[package]
name = "gdnative"
authors = ["The godot-rust developers"]
description = "The Godot game engine's gdnative bindings."
documentation = "https://docs.rs/crate/gdnative"
repository = "https://github.com/godot-rust/godot-rust"
homepage = "https://godot-rust.github.io/"
version = "0.9.3"
license = "MIT"
workspace = ".."
readme = "../README.md"
edition = "2018"

[features]
default = ["bindings"]
formatted = ["gdnative-bindings/formatted", "gdnative-bindings/one_class_one_file"]

gd_test = ["gdnative-core/gd_test"]
type_tag_fallback = ["gdnative-core/type_tag_fallback"]
bindings = ["gdnative-bindings"]

[dependencies]
gdnative-derive = { optional = true, path = "../gdnative-derive", version = "=0.9.3" }
gdnative-core = { optional = true, path = "../gdnative-core", version = "=0.9.3" }
gdnative-bindings = { optional = true, path = "../gdnative-bindings", version = "=0.9.3" }

[dev-dependencies]
trybuild = "1.0"
rustversion = "1.0"

[package.metadata.docs.rs]
all-features = true
