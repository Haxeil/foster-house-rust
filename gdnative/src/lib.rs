//! # Rust bindings for the Godot game engine
//!
//! This crate contains high-level wrappers around the Godot game engine's gdnative API.
//! Some of the types were automatically generated from the engine's JSON API description,
//! and some other types are hand made wrappers around the core C types.
//!
//! ## Memory management for core types
//!
//! Wrappers for most core types expose safe Rust interfaces, and it's unnecessary to mind
//! memory management most of the times. The exceptions are `VariantArray` and `Dictionary`,
//! internally reference-counted collections with "interior mutability" in Rust parlance. These
//! types are modelled using the "typestate" pattern to enforce that the official
//! [thread-safety guidelines][thread-safety]. For more information, read the type-level
//! documentation for these types.
//!
//! Since it is easy to expect containers and other types to allocate a copy of their
//! content when using the `Clone` trait, some types do not implement `Clone` and instead
//! implement [`NewRef`](./trait.NewRef.html) which provides a `new_ref(&self) -> Self` method
//! to create references to the same collection or object.
//!
//! ## Generated API types
//!
//! The `api` module contains high-level wrappers for all the API types generated from a
//! JSON description of the API. The generated types are tied to a specific version, which
//! is currently `3.2.3-stable` for the crates.io version. If you want to use the bindings
//! with another version of the engine, see the instructions [here][custom-version] on
//! generating custom bindings.
//!
//! ### Memory management
//!
//! API types may be reference-counted or manually-managed. This is indicated by the
//! `RefCounted` and `ManuallyManaged` marker traits.
//!
//! The API types can exist in three reference forms: bare, `TRef` and `Ref`. Bare references
//! to API types, like `&'a Node`, represent valid and safe references to Godot objects.
//! As such, API methods may be called safely on them. `TRef` adds typestate tracking, which
//! enable additional abilities like being able to be passed to the engine. `Ref`, or
//! "persistent" references, have `'static` lifetime, but are not always safe to use. For more
//! information on how to use persistent references safely, see the type-level documentation
//! on `Ref`.
//!
//! ## Feature flags
//!
//! ### `bindings`
//!
//! *Enabled* by default. Includes the crates.io version of the bindings in the `api` module.
//!
//! [thread-safety]: https://docs.godotengine.org/en/stable/tutorials/threads/thread_safe_apis.html
//! [custom-version]: https://github.com/godot-rust/godot-rust/#other-versions-or-custom-builds
//!

// TODO: add logo using #![doc(html_logo_url = "https://<url>")]

// Workaround: rustdoc currently shows hidden items in the original crate when they are
//             re-exported. Manually re-exporting the public items works around that.
#[doc(inline)]
pub use gdnative_core::{
    core_types, godot_dbg, godot_error, godot_gdnative_init, godot_gdnative_terminate, godot_init,
    godot_nativescript_init, godot_print, godot_warn, godot_wrap_method, nativescript, object,
    ref_kind, thread_access, GodotObject, GodotResult, NewRef, Null, Ref, TRef,
};

#[doc(hidden)]
pub use gdnative_core::*;

#[doc(inline)]
pub use gdnative_derive::*;

/// Curated re-exports of common items.
pub mod prelude;

#[doc(inline)]
#[cfg(feature = "bindings")]
/// Bindings for the Godot Class API.
pub use gdnative_bindings as api;
