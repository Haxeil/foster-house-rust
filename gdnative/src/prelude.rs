pub use gdnative_core::core_types::{
    self, error::GodotError, Aabb, Basis, ByteArray, Color, ColorArray, Dictionary, Float32Array,
    GodotString, Int32Array, NodePath, Plane, Quat, Rect2, Rid, StringArray, StringName, Transform,
    Transform2D, TypedArray, Variant, VariantArray, VariantDispatch, VariantOperator, VariantType,
    Vector2, Vector2Array, Vector3, Vector3Array,
};
pub use gdnative_core::core_types::{
    FromVariant, FromVariantError, OwnedToVariant, ToVariant, ToVariantEq,
};

pub use gdnative_core::object::{
    AsArg, GodotObject, Instanciable, Null, QueueFree, Ref, SubClass, TRef,
};
pub use gdnative_core::ref_kind::{ManuallyManaged, RefCounted};
pub use gdnative_core::thread_access::{Shared, ThreadLocal, Unique};
pub use gdnative_core::NewRef;

pub use gdnative_core::nativescript::{
    self,
    class::{Instance, RefInstance},
    init::{ClassBuilder, InitHandle, Method, MethodBuilder, Signal, SignalArgument},
    user_data::{self, Aether, ArcData, LocalCellData, MutexData, RwLockData},
    ExportInfo, NativeClass, NativeClassMethods, PropertyUsage,
};

pub use gdnative_core::{
    godot_dbg, godot_error, godot_gdnative_init, godot_gdnative_terminate, godot_init,
    godot_nativescript_init, godot_print, godot_site, godot_warn,
};

pub use gdnative_derive::*;

#[cfg(feature = "bindings")]
pub use gdnative_bindings::{
    Button, CanvasItem, CanvasLayer, ColorRect, Control, Image, Input, InputEvent, InputEventKey,
    KinematicBody, KinematicBody2D, Label, Node, Node2D, Object, PackedScene, Reference,
    ResourceLoader, SceneTree, Shader, Spatial, Sprite, Texture, Timer, Tween, Viewport,
};

#[cfg(feature = "bindings")]
pub use gdnative_bindings::utils::*;
