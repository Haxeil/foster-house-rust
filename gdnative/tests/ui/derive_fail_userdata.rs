use gdnative::prelude::*;

#[derive(NativeClass)]
#[inherit(Node)]
#[user_data]
struct Foo {}

impl Foo {
    fn new(_owner: &Node) -> Self {
        Foo {}
    }
}

fn main() {}
