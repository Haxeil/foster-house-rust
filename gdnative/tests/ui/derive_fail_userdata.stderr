error: expected attribute arguments in parentheses: #[user_data(...)]
 --> $DIR/derive_fail_userdata.rs:5:1
  |
5 | #[user_data]
  | ^^^^^^^^^^^^
